### PR TITLE
GSdx: WIP half bottom fix (Texture Shuffle Case)

### PR DESCRIPTION
**Pull request by tadanokojin**

### CODE REVIEW

File: plugins/GSdx/Renderers/HW/GSRendererHW.cpp [hunk 1]
```diff
@@ -523,6 +523,10 @@ void GSRendererHW::ConvertSpriteTextureShuffle(bool& write_ba, bool& read_ba)
 
 GSVector4 GSRendererHW::RealignTargetTextureCoordinate(const GSTextureCache::Source* tex)
 {
+	if (tex->m_is_middle)
+	{
+		return GSVector4(0, 3584);
```
Comment (tadanokojin) [inline plugins/GSdx/Renderers/HW/GSRendererHW.cpp]: Woops, I did positive here to test but in screenshot I posted to discord it was minus.

**Comment by legend800** [comment]
No change in Path of Neo. Darn, been waiting for this fix for years. :)

PR:
2019-04-21 09_32_26-Slot_ 0 _ Speed_ 100% (59 95) _ Limiter_ Normal _ GSdx OGL HW _ 2560x2240 _ Auto

Native Res.
2019-04-21 09_33_13-Slot_ 0 _ Speed_ 100% (60 01) _ Limiter_ Normal _ GSdx OGL SW _ 512x448 _ Automa

**Comment by MrCK1** [comment]
Just remember it's only a WIP/Proof of Concept. Give it some time to improve, I wouldn't expect anything changes yet.

**Comment by legend800** [comment]
Matrix - Path of Neo fixed on latest commit. Amazing work!

2019-04-22 08_21_33-Window

2019-04-22 08_25_19-Slot_ 0 _ Speed_ 100% (59 98) _ Limiter_ Normal _ GSdx OGL HW _ 2560x2240 _ Auto

**Comment by Silanda** [comment]
Need for Speed: Hot Pursuit 2's improved as well. The shadow on the road is gone, but the busted scenes are broken:

gsdx_20190422164046

**Comment by tadanokojin** [comment]
@Silanda was busted screen broken on master or is this new?
I updated the detection maybe it will fix it.

**Comment by Silanda** [comment]
@tadanokojin My bad. It looks like it may have been bust in master. I'd been playing in software mode and hadn't checked.

Also the change on line 452 of GSRendererHW.cpp from `bool half_bottom = height_delta <= 5;` to `bool half_bottom = abs(height_delta) <= 5;` causes the graphics glitch to return.

**Comment by tadanokojin** [comment]
Can you send a gsdump of the issue for me?
https://forums.pcsx2.net/Thread-How-to-create-a-proper-GS-dump

Couple of frames please.

**Comment by Silanda** [comment]
Here are a couple of single frame dumps from in game.
[NFSHP2Dumps.zip](https://github.com/PCSX2/pcsx2/files/3105077/NFSHP2Dumps.zip)

**Comment by JOHN30011887** [comment]
How can i download this to test for crash twinsanity, sorry dont know what/where to click

**Comment by MrCK1** [comment]
https://ci.appveyor.com/project/gregory38/pcsx2/builds/24035294/job/42m8jmv990u7ywem/artifacts

File: plugins/GSdx/Renderers/HW/GSRendererHW.cpp [hunk 2]
```diff
@@ -448,6 +452,9 @@ void GSRendererHW::ConvertSpriteTextureShuffle(bool& write_ba, bool& read_ba)
 	tex_pos &= 0xFF;
 	read_ba = (tex_pos > 112 && tex_pos < 144);
 
+	int height_delta = m_src->m_valid_rect.height() - m_r.height();
+	bool half_bottom = abs(height_delta) <= 64 && !m_disable_ts_half_bottom;;
```
Comment (tadanokojin) [inline plugins/GSdx/Renderers/HW/GSRendererHW.cpp]: note to self to remove extra `;` when I rebase

**Comment by tadanokojin** [comment]
@Silanda it should be fixed

**Comment by Silanda** [comment]
Looks good here!

**Comment by JOHN30011887** [comment]
So i tested Crash Twinsanity since it was marked as fixed but sorry to be that guy but its not :(
It is alot better than what it was though, as before the top half of screen was yellow fog.

But now theres a yellow line where the bottom of the fog wall would be
pcsx2_2019_04_23_23_02_55_897

Doesnt make any difference on settings, happens in dx11 and opengl
only time it was gone was using ps2 resolution option, anything higher the solid yellow line appears again

**Comment by tadanokojin** [comment]
@JOHN30011887 yes, it is an upscaling issue unrelated to half bottom issue. You can try maybe a TC offset.

**Comment by JOHN30011887** [comment]
oh, so half bottom issue is indeed fixed but now its another issue, ah ok then, thanks for info, il mess with that setting and see :)

**Comment by tadanokojin** [comment]
@JOHN30011887 how about half pixel offset special (texture - aggressive)

**Comment by JOHN30011887** [comment]
Tried that setting, the line is alot thiner so is harder to see but still there and also gives slight corruption to crash, he ends up with thin red lines on him when that setting in on texture - aggressive

**Comment by atomic83GitHub** [comment]
Try half pixel offset to specia-texture instead and put round sprite at half, this usualy work for games that have this line such as nemo or just cause.

**Comment by Xcedf** [comment]
@JOHN30011887 use Half-pixel Offset Special (Texture-Aggressive) setting
See for youself
4xIR Normal-Vertex offset setting

Special-Texture

and Texture_Aggressive

**Comment by tadanokojin** [comment]
Yes but did you try special-texture **with** round sprite at half like atomic asked?

**Comment by atomic83GitHub** [comment]
Yeah, I will remove my comments in this thread.

**Comment by JOHN30011887** [comment]
@atomic83GitHub, tried that just now but no luck, still a very thin line, harder to see but still there, the line flickers abit
@Xcedf, i did already and the line still shows in all your pics

**Comment by denoid24** [comment]
> Matrix - Path of Neo fixed on latest commit. Amazing work!
> 
> 2019-04-22 08_21_33-Window
> 
> 2019-04-22 08_25_19-Slot_ 0 _ Speed_ 100% (59 98) _ Limiter_ Normal _ GSdx OGL HW _ 2560x2240 _ Auto

Hey) i'm a new one noob here on github, please tell me how to apply the fix to my pcsx2) would be infinitely grateful for the help)

**Comment by tadanokojin** [comment]
@denoid24 
You don't need to apply it. You can download and setup a dev build.
Assisting you with that is not what this thread is for.

Please go to https://forums.pcsx2.net/ create an account and ask there if you need help.
Alternatively you can ask on discord.

**Comment by denoid24** [comment]
Thanks) so all that's needed is to download the latest dev build?)

пн, 29 апр. 2019 г. в 21:04, Kojin <notifications@github.com>:

**Comment by denoid24** [comment]
Just downloaded the latest devbuild, no changes still as before. did i pick
the wrong build?
[image: image.png]

пн, 29 апр. 2019 г. в 22:57, Денис Карасёв <denoid2@gmail.com>:

**Comment by legend800** [comment]
Stop spamming this PR. Take it to the forums to learn how to test pcsx2.

**Pull request merged**